src/lua/CommunityBalanceMod/FortressPvE/Whip.lua

```diff
@@ -0,0 +1,262 @@
+    function Whip:UpdateResearch()
+
+        local researchId = self:GetResearchingId()
+
+        if researchId == kTechId.UpgradeToFortressWhip then
+        
+            local techTree = self:GetTeam():GetTechTree()    
+            local researchNode = techTree:GetTechNode(kTechId.Whip)   -- get a progress bar at the Whip in the tech tree. TODO Does this affect spec, comm view?
+            researchNode:SetResearchProgress(self.researchProgress)
+            techTree:SetTechNodeChanged(researchNode, string.format("researchProgress = %.2f", self.researchProgress)) 
+            
+        end
+
+    end
+
+
+    function Whip:OnResearchCancel(researchId)
+
+        if researchId == kTechId.UpgradeToFortressWhip then
+        
+            local team = self:GetTeam()
+            
+            if team then
+            
+                local techTree = team:GetTechTree()
+                local researchNode = techTree:GetTechNode(kTechId.Whip)
+                if researchNode then
+                    researchNode:ClearResearching()
+                    techTree:SetTechNodeChanged(researchNode, string.format("researchProgress = %.2f", 0))   
+                end
+            end  
+        end
+    end
+
+    -- Called when research or upgrade complete
+    function Whip:OnResearchComplete(researchId)
+
+        if researchId == kTechId.UpgradeToFortressWhip then
+        
+            --self:SetTechId(kTechId.FortressWhip)
+            self:UpgradeToTechId(kTechId.FortressWhip)
+            
+            local techTree = self:GetTeam():GetTechTree()
+            local researchNode = techTree:GetTechNode(kTechId.Whip)
+            
+            if researchNode then     
+    
+                researchNode:SetResearchProgress(1)
+                techTree:SetTechNodeChanged(researchNode, string.format("researchProgress = %.2f", self.researchProgress))
+                researchNode:SetResearched(true)
+                techTree:QueueOnResearchComplete(kTechId.FortressWhip, self)
+
+                
+
+                
+            end
+            
+        end
+    end
+
+end
+
+
+if Client then
+    
+    function Whip:OnUpdateRender()
+
+           if not self.fortressWhipMaterial and self:GetTechId() == kTechId.FortressWhip then
+ 
+                local model = self:GetRenderModel()
+
+                if model and model:GetReadyForOverrideMaterials() then
+                
+                    model:ClearOverrideMaterials()
+                    --local material = GetPrecachedCosmeticMaterial( "Whip", "Fortress" )
+                    local material = Whip.kfortressWhipMaterial
+                    assert(material)
+                    model:SetOverrideMaterial( 0, material )
+
+                    self.fortressWhipMaterial = true
+                end
+
+           end
+    end
+end
+
+
+function Whip:OnAdjustModelCoords(modelCoords)
+    --gets called a ton each second
+
+    if not self.reduceSized and self:GetTechId() == kTechId.Whip then
```

Comment:
Is this supposed to be self.reduce**d**Sized?

Reply:
typo, should be right now